app/worker/utils.go

```diff
@@ -15,17 +15,57 @@
 package worker
 
 import (
-	"fmt"
+	"os"
+	"regexp"
 	"strings"
 
 	"github.com/mendersoftware/workflows/model"
+	"github.com/thedevsaddam/gojsonq"
 )
 
+const (
+	workflowEnvVariable   = "env."
+	workflowInputVariable = "workflow.input."
+)
+
+var reExpression = regexp.MustCompile(`\$\{([^\}]+)\}`)
+var reExpressionOutput = regexp.MustCompile(`(.*)\.output\.(.*)`)
+
 func processJobString(data string, workflow *model.Workflow, job *model.Job) string {
-	for _, param := range job.InputParameters {
-		data = strings.ReplaceAll(data,
-			fmt.Sprintf("${workflow.input.%s}", param.Name),
-			param.Value)
+	matches := reExpression.FindAllStringSubmatch(data, -1)
+
+	for _, submatch := range matches {
+		match := submatch[1]
```

Comment:
should it be `submatch[0]` here?
I would add comment here: first match is the variable identifier to replace

Reply:
No, `submatch[0]` is the whole string that matched, and `submatch[1]` is the first matching group.